test/contrib/test_module.py

```diff
@@ -311,3 +316,902 @@ def model():
+    deep_params = {
+        "nested.deep_nested.weight": jnp.array([[9.0, 10.0], [11.0, 12.0]]),
+        "nested.deep_nested.bias": jnp.array([0.5, 0.6]),
+    }
+
+    updated_state = _update_state_with_params(state, deep_params)
+
+    # Check deep nested parameters
+    assert jnp.array_equal(
+        updated_state["nested"]["deep_nested"]["weight"].value,
+        deep_params["nested.deep_nested.weight"],
+    )
+    assert jnp.array_equal(
+        updated_state["nested"]["deep_nested"]["bias"].value,
+        deep_params["nested.deep_nested.bias"],
+    )
+
+    # Test Case 3: Mixed parameter update (flat and nested)
+    mixed_params = {
+        "weight": jnp.array([[13.0, 14.0], [15.0, 16.0]]),
+        "nested.deep_nested.bias": jnp.array([0.7, 0.8]),
+    }
+
+    updated_state = _update_state_with_params(state, mixed_params)
+
+    # Check mixed parameters
+    assert jnp.array_equal(updated_state["weight"].value, mixed_params["weight"])
+    assert jnp.array_equal(
+        updated_state["nested"]["deep_nested"]["bias"].value,
+        mixed_params["nested.deep_nested.bias"],
+    )
+
+    # Test Case 4: Non-existent parameters (should not raise errors)
+    nonexistent_params = {
+        "nonexistent": jnp.array([1.0, 2.0]),
+        "nested.nonexistent": jnp.array([3.0, 4.0]),
+        "nonexistent.nested": jnp.array([5.0, 6.0]),
+    }
+
+    # This should not raise errors
+    updated_state = _update_state_with_params(state, nonexistent_params)
+
+    # Original parameters should remain unchanged
+    assert jnp.array_equal(updated_state["weight"].value, mixed_params["weight"])
+    assert jnp.array_equal(
+        updated_state["nested"]["deep_nested"]["bias"].value,
+        mixed_params["nested.deep_nested.bias"],
+    )
+
+    # Test Case 5: Empty parameters dictionary
+    empty_params = {}
+
+    updated_state = _update_state_with_params(state, empty_params)
+
+    # State should remain unchanged
+    assert jnp.array_equal(updated_state["weight"].value, mixed_params["weight"])
+    assert jnp.array_equal(
+        updated_state["nested"]["deep_nested"]["bias"].value,
+        mixed_params["nested.deep_nested.bias"],
+    )
+
+
+def test_nnx_module_model_surgery():
+    """Test the model surgery approach used in the nnx_module function."""
+    from flax import nnx
+    import jax.numpy as jnp
+
+    import numpyro
+    from numpyro import handlers
+    from numpyro.contrib.module import nnx_module
+
+    # Create a simple module with parameters
+    class SimpleModule(nnx.Module):
+        def __init__(self, din, dout, *, rngs):
+            self.weight = nnx.Param(jnp.zeros((din, dout)))
+            self.bias = nnx.Param(jnp.zeros(dout))
+
+        def __call__(self, x):
+            return x @ self.weight + self.bias
+
+    # Define a model that uses the module
+    def model(x, y=None):
+        # Create the module
+        nn = nnx_module("nn", SimpleModule, din=3, dout=1)
```

Comment:
I think `nnx` encourages "eager initialization" (its main selling point). So it is better to initialize the module outside of numpyro `model`. Something like
```
module = SimpleModule(...)

def model(x, y=None):
    nn = nnx_module("nn", module)
```
With nnx, you can inspect the module directly so lots of code can be simplified. For example, you can do
```
_, params, state = nnx.split(module, nnx.Param, nnx.filterlib.to_predicate(nnx.Not(nnx.Param)))
```
to extract params and mutable state (which stores rngs, so we don't need numpyro to pass rng to it, yay!)

Reply:
yes! added this!